apps/src/weblab/CdoBramble.js

```diff
@@ -202,6 +216,49 @@ export default class CdoBramble {
     getStore().dispatch(actions.changeProjectSize(bytes));
   }
 
+  uploadAllFilesToServer(callback) {
+    this.shell().ls(this.projectPath, (err, entries) => {
+      if (err) {
+        console.error(
+          `CdoBramble failed to receive file entries from Bramble. ${err}`
+        );
+        callback(err);
+      }
+
+      const uploadEntry = index => {
+        const next = (err, newVersionId) => {
+          if (err) {
+            callback(err);
+            return;
+          }
+
+          this.lastSyncedVersionId = newVersionId;
+          if (index >= entries.length - 1) {
+            callback(null, true /* preWriteHook was successful */);
+          } else {
+            uploadEntry(index + 1);
+          }
+        };
+
+        const filename = entries[index].path; // 'path' is relative, so it will be the filename.
+        this.getFileData(this.prependProjectPath(filename), (err, fileData) => {
+          if (err) {
+            callback(err);
+          } else {
+            this.api.changeProjectFile(
+              filename,
+              fileData,
+              next,
+              true /* skipPreWriteHook because we are calling from the preWriteHook */
+            );
+          }
+        });
+      };
+
+      uploadEntry(0);
```

Comment:
I don't think I'm understanding why this line is here?

Reply:
this starts the recursive process -- calling `uploadEntry(0)` will upload the file at index 0, and `uploadEntry` will handle calling itself until all indices have been uploaded